File: KotlinStudy/step03_listvuew/src/main/java/com/example/step03_listvuew/MainActivity.java
package com.example.step03_listvuew;

import android.app.AlertDialog;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity
        implements AdapterView.OnItemClickListener, AdapterView.OnItemLongClickListener {
    //필요한 필드 선언
    ArrayAdapter<String> adapter;
    List<String> names;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);

        //ListView 에 출력할 sample 데이터
        names =new ArrayList<>();
        names.add("ezfz");
        names.add("hysz");
        for(int i=0; i<100; i++){
            names.add("ezfz"+i);
        }
        //new ArrayAdapter<>(Context type, 셀의 레이아웃 정보, 모델)
        //ListView 에 연결할 Adapter 객체 생성하기
        adapter=new ArrayAdapter<>(this,
                android.R.layout.simple_list_item_1, names);
        //ListView 의 참조값
        ListView listView=findViewById(R.id.listView);
        //ListView 에 아답타 연결하기
        listView.setAdapter(adapter);
        //ListView 에 아이템 클릭 리스너 등록
        listView.setOnItemClickListener(this);
        listView.setOnItemLongClickListener(this); // 꾹 누르는 경우

        // ### add ###

        EditText inputName = findViewById(R.id.inputName);
        Button addBtn = findViewById(R.id.addBtn);

        addBtn.setOnClickListener(view -> {
            names.add(inputName.getText().toString());
            inputName.setText("");
            adapter.notifyDataSetChanged();
        });



    }


    //AdapterView.OnItemClickListener 인터페이스를 구현해서 재정의한 메소드
    @Override
    public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
        // view 에는 cell View 의 참조값이 들어 있고, i 는 선택한 cell 의 인덱스가 들어 있다.

        //선택한 cell 에 출력된 문자열 얻어내기
        String selected = adapter.getItem(i);

        Toast.makeText(this, selected+" 가 선택됨!", Toast.LENGTH_SHORT).show();
    }

    @Override
    public boolean onItemLongClick(AdapterView<?> adapterView, View view, int i, long l) {

        new AlertDialog.Builder(this)
                .setTitle("알림")
                .setMessage("삭제 기?")
                .setPositiveButton("네", (dialogInterface, i1) -> {
                    names.remove(i);
                    // 삭제하는 것 만으로는 화면이 업데이트가 안됨 -> 변경 알림 -> ui update
                    adapter.notifyDataSetChanged();
                })
                .setNegativeButton("아니요", null)
                .create()
                .show();

        return false;
    }
}

File: KotlinStudy/step07customlistview/src/main/java/com/example/step07customlistview/MainActivity.kt
package com.example.step07customlistview

import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.EditText
import android.widget.ListView
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

/*
    [  json 문자열 사용하기 ]
    [ ] => json array
    {} => json object
 */
class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // model 객체
        val list:MutableList<MemberDto> = mutableListOf()

        // ListView에 연결할 어댑터객체 생성해서 연결하기
        val adapter = MemberAdapter(this, R.layout.listview_cell, list)

        val listView:ListView = findViewById(R.id.listView)
        listView.adapter = adapter

        val inputName:EditText = findViewById(R.id.inputName)
        val inputAddr:EditText = findViewById(R.id.inputAddr)

        val addBtn:Button = findViewById(R.id.addBtn)

        addBtn.setOnClickListener{
            lifecycleScope.launch(Dispatchers.Main) {
                addMember(inputName.text.toString(), inputAddr.text.toString())
            }


        }

        // json 문자열을 log에 출력
        lifecycleScope.launch(Dispatchers.Main) {
            val result = getMemeber()
            Log.d("### result ### ", result)
            // 응답받은 문자열의 형식이 [{},{},{}, ...] -> JSONArray 객체 생성
            val array = JSONArray(result)

            for (i in 0 until array.length()){
                val cur = array.getJSONObject(i)
//                list.add(MemberDto(cur.getInt("num"), cur.getString("name"), cur.getString("addr")))
                val mem = MemberDto().apply {
                    num = cur.getInt("num")
                    name = cur.getString("name")
                    addr = cur.getString("addr")
                }
                list.add(mem)
            }
            adapter.notifyDataSetChanged()
        }
    }

    suspend fun getMemeber():String {
        val result = withContext(Dispatchers.IO){

            val builder = StringBuilder()

            try {
                //요청 url 을 생성자의 인자로 전달하면서 URL 객체를 생성한다
                val url = URL("http://192.168.0.205:8888/api/members")
                //URLConnection 객체를 원래 type (자식 type) 으로 casting 해서 받는다.
                val conn = url.openConnection() as HttpURLConnection
                if (conn != null) {
                    conn.connectTimeout = 20000 //응답을 기다리는 최대 대기 시간
                    conn.requestMethod = "GET" // 요청 메소드 설정 (Default 는 GET)
                    conn.useCaches = false //케쉬 사용 여부
                    //응답 코드를 읽어온다.
                    val responseCode = conn.responseCode
                    if (responseCode == HttpURLConnection.HTTP_OK) { //정상 응답이라면 (200)
                        //문자열을 읽어들일수 있는 객체의 참조값 얻어오기
                        val br =
                            BufferedReader(InputStreamReader(conn.inputStream))
                        //반복문 돌면서
                        while (true) {
                            //문자열을 한줄씩 읽어 들인다.
                            val line = br.readLine() ?: break
                            //StringBuilder 객체에 누적 시키기
                            builder.append(line)
                        }
                    }
                }
            } catch (e: Exception) {
                Log.e("MainActivity", e.message!!)
            } //try~catch

            builder.toString()
        }
        return result
    }

    suspend fun addMember(name:String, addr:String):String {

        val result = withContext(Dispatchers.IO){

            val builder = StringBuilder()

            try {
                //요청 url 을 생성자의 인자로 전달하면서 URL 객체를 생성한다
                val url = URL("http://192.168.0.205:8888/api/members")
                //URLConnection 객체를 원래 type (자식 type) 으로 casting 해서 받는다.
                val conn = url.openConnection() as HttpURLConnection
                if (conn != null) {
                    conn.connectTimeout = 20000 //응답을 기다리는 최대 대기 시간
                    conn.requestMethod = "POST" // 요청 메소드 설정 (Default 는 GET)
                    conn.useCaches = false //케쉬 사용 여부
                    conn.setRequestProperty("Content-Type", "application/json")
                    //문자열을 출력하기 위한 객체
                    val osw = OutputStreamWriter(conn.outputStream)
                    osw.write("""
                        { 
                            "name" : $name,
                            "addr" : $addr
                        }
                    """.trimIndent())
                    osw.flush()

                    //응답 코드를 읽어온다.
                    val responseCode = conn.responseCode
                    if (responseCode == HttpURLConnection.HTTP_OK) { //정상 응답이라면 (200)
                        //문자열을 읽어들일수 있는 객체의 참조값 얻어오기
                        val br =
                            BufferedReader(InputStreamReader(conn.inputStream))
                        //반복문 돌면서
                        while (true) {
                            //문자열을 한줄씩 읽어 들인다.
                            val line = br.readLine() ?: break
                            //StringBuilder 객체에 누적 시키기
                            builder.append(line)
                        }
                    }
                }
            } catch (e: Exception) {
                Log.e("MainActivity", e.message!!)
            } //try~catch

            builder.toString()
        }
        return result
    }
}